src/ray/util/subreaper.cc

```diff
@@ -0,0 +1,177 @@
+        handler(error, signal_number);
+        RegisterSignalHandlerLoop(signals, handler);
+      });
+}
+
+void SigchldHandlerReapZombieAndRemoveKnownChildren(
+    const boost::system::error_code &error, int signal_number) {
+  if (error) {
+    RAY_LOG(WARNING) << "Error in SIGCHLD handler: " << error.message();
+  }
+  int status;
+  pid_t pid;
+  // Reaps any children that have exited. WNOHANG makes waitpid non-blocking and returns
+  // 0 if there's no zombie children.
+  while ((pid = waitpid(-1, &status, WNOHANG)) > 0) {
+    if (WIFEXITED(status)) {
+      RAY_LOG(INFO) << "Child process " << pid << " exited with status "
+                    << WEXITSTATUS(status);
+    } else if (WIFSIGNALED(status)) {
+      RAY_LOG(INFO) << "Child process " << pid << " exited from signal "
+                    << WTERMSIG(status);
+    }
+    KnownChildrenTracker::instance().RemoveKnownChild(pid);
+  }
+}
+
+}  // namespace
+
+// Set this process as a subreaper.
+// Only works on Linux >= 3.4.
+bool SetThisProcessAsSubreaper() {
+  if (prctl(PR_SET_CHILD_SUBREAPER, 1) == -1) {
+    RAY_LOG(WARNING) << "Failed to set this process as subreaper: " << strerror(errno);
+    return false;
+  }
+  return true;
+}
+
+void SetSigchldIgnore() { signal(SIGCHLD, SIG_IGN); }
+
+void SetupSigchldHandlerRemoveKnownChildren(boost::asio::io_context &io_service) {
+  auto sigchld_signals = std::make_shared<boost::asio::signal_set>(io_service, SIGCHLD);
+  RegisterSignalHandlerLoop(sigchld_signals,
+                            SigchldHandlerReapZombieAndRemoveKnownChildren);
+  RAY_LOG(INFO)
+      << "Raylet is set to reap zombie children and remove known children pids.";
+}
+
+// Kill all child processes that are not owned by this process.
+// It's done by checking the list of child processes and killing the ones that are not
+// created via Process::spawnvpe.
+//
+// TODO: Checking PIDs is not 100% reliable because of PID recycling. If we find issues
+// later due to this, we can use pidfd.
+void KillUnknownChildren() {
+  auto to_kill =
+      KnownChildrenTracker::instance().ListUnknownChildren([]() -> std::vector<pid_t> {
+        auto child_procs = GetAllProcsWithPpid(GetPID());
+        if (!child_procs) {
+          RAY_LOG(FATAL) << "Killing leaked procs not supported on this platform. Only "
+                            "supports Linux >= 3.4";
+          return {};
+        }
+        return *child_procs;
+      });
+  for (auto pid : to_kill) {
+    RAY_LOG(INFO) << "Killing leaked child process " << pid;
+    auto error = KillProc(pid);
+    if (error && (*error)) {
+      RAY_LOG(WARNING) << "Failed to kill leaked child process " << pid << " with error "
+                       << error->message() << ", value = " << error->value();
+    }
+  }
+}
+
+void KnownChildrenTracker::AddKnownChild(std::function<pid_t()> create_child_fn) {
+  if (!enabled_) {
+    create_child_fn();
+    return;
+  }
+  absl::MutexLock lock(&m_);
```

Comment:
we don't need a lock when we create a child proc? 

```
pid_t pid = create_child_fn();
absl::MutexLock lock(&m_);
children_.insert(pid);
```

Reply:
We don't want the creation of a pid (in procfs) happen when we are reading the procfs for pids, I assume

Reply:
e.g. if 2 racing threads:

1. create process
    a. then add pid to the known list
3. kill unknown children

If the order is `1 -> 3 -> a`, the new process is killed which is not what we want. We have to make `1` and `a` be atomic on the list.

Reply:
see the race condition I gave: if a killing happens between `create()` and `addknownchild` it may kill the newborn child process.